File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() < 2) return s.length();
        int max = 0;
        char[] split = s.toCharArray();
        int[] map = new int[256];
        for(int i = 0, j = 0; i < split.length; i++) {
            char c = split[i]; 
            j = Math.max(j, map[c]);
            map[c] = i + 1;
            max = Math.max(max, i - j + 1);
        }
        return max;
    }
}

File: 3. Longest Substring Without Repeating Characters.js
// LEETCODE
// Given a string s, find the length of the longest substring without repeating characters.

// -------------------------------------------------------------------------------------------  
// Input: s = "abcabcbb"
// Output: 3
// Explanation: The answer is "abc", with the length of 3.

// -------------------------------------------------------------------------------------------

var lengthOfLongestSubstring = function(s) {
    let start=0;
    let end=0;
    let maxLength=0;
    let uniqueChar=new Set();
    while(end<s.length){
        if(!uniqueChar.has(s[end])){
            uniqueChar.add(s[end])
            end++
            maxLength=Math.max(maxLength,uniqueChar.size)
        } else {
            uniqueChar.delete(s[start])
            start++
        }
    }
    return maxLength;
};

// time complexicity => O(n)
